### DISCUSSION TITLE
Feature request: Support pick Intent for multiple contacts?

### DISCUSSION THREAD

**Discussion by tinacious** [vestrel00/contacts-android#346] [Ideas]
First I'd like to say thank you for creating this amazing library. It's been a huge time-saver and has allowed me to get relevant fields for contacts quite easily without having to fumble with the `contentResolver` API's.

I was wondering if it would be possible to support a pick intent for multiple contacts.

Reason: Many users do not feel comfortable giving full access to all their contacts.

Example UX: 

- User taps a button
- I launch an intent for an activity result (perhaps your library can craft this intent)
- I can provide the returned data from the activity result to your library to get contacts (e.g. the same data type that would be returned from `Contacts(context).query().find()`, e.g. `Query.Result` or `List<Contact>`.

This is some code I have for picking a single contact. It doesn't get the fields I need (that your library easily provides) like first name and last name, and it doesn't support multiple contacts, which makes the UX quite cumbersome:

```kt
private fun launchPickOneContactIntent() {
    val contactPickIntent = Intent(
        Intent.ACTION_PICK,
        ContactsContract.Contacts.CONTENT_URI
    ).also {
        it.type = ContactsContract.CommonDataKinds.Phone.CONTENT_TYPE
    }
    pickSingleContact.launch(contactPickIntent)
}

private val pickSingleContact = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
    val uri = result.data?.data ?: return@registerForActivityResult
    Logger.d("URI: $uri")

    val pickedContact = viewModel.getContactFromPickIntentUri(uri) // this uses the contentResolver APIs and only gets the phone number and display name (not enough data)
    Logger.d("Picked contact: $pickedContact")
}
```

Is it possible to:

- pick multiple contacts with implicit intents that do not require elevated privileges? I'm not even sure if Android allows this anymore—I have seen some StackOverflow posts that claim this is possible but they are quite old and those snippets did not work on the newer versions I'm testing on
- get all relevant contact data that your library provides (e.g. first name, last name, email, phone number, normalized phone number, etc.) for the contacts provided with the pick intent

Let me know if I can provide any additional information as I'd love to be able to support users who do not want to share their full contacts list.

Thank you! 🙏

**Comment by vestrel00** [comment]
Hey @tinacious! Glad to hear that this library has helped you out 😁 

> pick multiple contacts with implicit intents that do not require elevated privileges?

AFAIK, there is no intent for picking multiple contacts. The only thing I see in the [official docs is for picking a single contact](https://developer.android.com/guide/components/intents-common#PickContact), which is exactly what you are doing in the code snippet you provided. 

> get all relevant contact data that your library provides

As long as you are able to get Contact or RawContact ID(s), the you can use the APIs provided in this library to get all of the data you need.

For example, if you have a list of Contact IDs, then getting all of the Contacts using this API would look like...

```kotlin
val contactIds = listOf<Long>();
val contacts = Contacts(context)
    .query()
    .where { Contact.Id `in` contactIds }
    .find()
```

If you have a list of RawContact IDs and you want the list of Contacts...

```kotlin
val rawContactIds = listOf<Long>();
val contacts = Contacts(context)
    .query()
    .where { RawContact.Id `in` rawContactIds }
    .find()
```

If instead you want to get RawContacts directly...

```kotlin
val rawContactIds = listOf<Long>();
val rawContacts = Contacts(context)
    .rawContactsQuery()
    .where { RawContact.Id `in` rawContactIds }
    .find()
```

> I was wondering if it would be possible to support a pick intent for multiple contacts.
> 
> Reason: Many users do not feel comfortable giving full access to all their contacts.

Hmm.. I'm a little bit confused here. In order to use the APIs provided in this library, your app must explicitly request the `android.permission.READ_CONTACTS` permission. This will give your app "full access to all their contacts".

In any case, your reasons are not my business so I won't stick my nose in it 😃 So, I'll focus on **"would be possible to support a pick intent for multiple contacts"**.

If Android itself does not provide a builtin intent to pick multiple contacts, then the app must implement an Activity to handle such an intent. There are several ways you can go about this.

1. Find another open source library that provides an Activity that allows you to pick multiple contacts. Then use the APIs in this library to get all contact data (assuming that the other open source library gives you access to the Contact ID).
2. Implement your own Activity that allows you to pick multiple contacts. You can use the APIs in this library to show all available contacts users can pick from.
3. I can implement an Activity and add it into this library, which you can then use.

I am hesitant to go with option 3 because the main purpose of this library is to provide CRUD APIs. I do have an experimental WIP `ui` module that could potentially contain option 3 in the future. However, I currently have no plans to do that for the time-being because work and life are busy and it is not part of the [v1.0.0 project roadmap](https://github.com/vestrel00/contacts-android/wiki/Project-Roadmap) I have planned out. It's also going to be tricky because styling such an Activity for general purpose use and giving users ways to customize the look-and-feel will take quite some time to do.

Therefore, I suggest that you take option 2 (or option 1).

-------

LMK if that helped answer some of your questions 🙏

**Comment by tinacious** [comment]
Thanks for the quick reply!

With regards to your comment: 

> Hmm.. I'm a little bit confused here. In order to use the APIs provided in this library, your app must explicitly request the android.permission.READ_CONTACTS permission. This will give your app "full access to all their contacts".

I was able to use your library on a device that granted contacts permissions and it works great! For a device that doesn't grant contacts permissions, I didn't try to use your library since I wasn't able to pick multiple contacts from Android. Your idea to provide the contact IDs though seems like it would be a good next step **if** I were able to get a list of picked contacts (which doesn't seem possible right now, at least officially). 

I was hoping that I could use a pick intent and get the data from the provided content URI, similar to how you can pick an image without requesting access to a user's entire photo gallery. In the case of images, I always get the relevant data I need. I don't know if that is the case with contacts.

> In any case, your reasons are not my business so I won't stick my nose in it 😃 

Happy to share again! I had mentioned I'd like to support users who are not comfortable sharing their contacts list:

> Reason: Many users do not feel comfortable giving full access to all their contacts.

I can, however, elaborate further – I fear users will see the contact permissions request and uninstall the app. I suspect there will be a significant amount of drop-off for security/privacy-focused users. I wanted to provide an alternate option where users can choose which contacts they're comfortable sharing instead of their entire address book, and at this time, sharing a contact one-by-one doesn't seem like a good user experience.

Do you know if your library would be able to support getting first name, last name and phone number from a single picked contact from the code I provided in the initial message? Or is contacts permissions required to query the referenced data for columns in `android.provider.ContactsContract.CommonDataKinds.StructuredName.*` (e.g. given name, family name, etc.). At a glance, it seems quite difficult to see which data is included with the officially-documented single contact pick intent, and I understand multiple queries with the `contentResolver` need to happen to get the additional fields (outside of phone number and display name) and I'm wondering if those require contacts permissions (since they appear to be in a different "table"). So I'm wondering if it's even possible, if I were to get the contact IDs (since that appears to be one of the available fields last I checked) if I could pass them to your library with the code snippets you provided (or if it would just crash when attempting to query related fields since it doesn't have permissions).

With regards to this:

> If Android itself does not provide a builtin intent to pick multiple contacts, then the app must implement an Activity to handle such an intent.

I'm already doing that for users who have granted contacts permissions (with the help of your library). My main goal is to support an alternate experience for users who do not grant this permission and I was hoping to do so via Android pick intents (similar to how we can allow a user to pick a photo from any app on their phone that supports it without needing full photo gallery access).

> I can implement an Activity and add it into this library, which you can then use. [...] I am hesitant to go with option 3 because the main purpose of this library is to provide CRUD APIs.

This is not necessary and I agree with you entirely. And even if you were to do that, I don't think it would solve the problem since it likely requires elevated permissions to scan all the contacts to display them.

Anyhow, hope that helps explain the reasons! Let me know if I can answer anything else!

**Reply by vestrel00** [reply]
> My main goal is to support an alternate experience for users who do not grant this permission (READ_CONTACTS)

Ahhh, got it! It makes more sense now. This part made everything clear to me.

> Do you know if your library would be able to support getting first name, last name and phone number from a single picked contact from the code I provided in the initial message? Or is contacts permissions required to query the referenced data for columns in android.provider.ContactsContract.CommonDataKinds.StructuredName.* (e.g. given name, family name, etc.)...(or if it would just crash when attempting to query related fields since it doesn't have permissions).

Unfortunately, my library requires contacts permission to be granted in order for any library functions to work properly. My library will not throw exceptions if permissions are not granted but it will return failed results (empty list, null, or false,).

I could implement utility functions that would take in activity result data and parse "from a single picked contact from the code (you) provided in the initial message" without querying the ContactsProvider. However, I would like to hold off on supporting such non-essential things like that until after v1.0.0.

So, you will have to extract the data you need from the activity result data yourself. I apologize for not being of much help 🙏
